runtime/common/src/xcm.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{traits::ProcessMessageError, weights::Weight};
+use xcm::v3::{
+	Instruction::{
+		self, BuyExecution, ClaimAsset, ClearOrigin, ReceiveTeleportedAsset, ReserveAssetDeposited,
+		WithdrawAsset,
+	},
+	MultiLocation,
+	WeightLimit::{Limited, Unlimited},
+};
+use xcm_builder::{CreateMatcher, MatchXcm};
+use xcm_executor::traits::ShouldExecute;
+
+/// Allows execution from all origins taking payment into account.
+///
+/// Only allows for `TeleportAsset`, `WithdrawAsset`, `ClaimAsset` and
+/// `ReserveAssetDeposit` XCMs because they are the only ones that place assets
+/// in the Holding Register to pay for execution. This is almost equal to
+/// [`xcm_builder::AllowTopLevelPaidExecutionFrom<T>`] except that it allows for
+/// multiple assets and is not generic to allow all origins.
+
+pub struct AllowTopLevelPaidExecution;
```

Comment:
I'm guessing this was removed from upstream and copied here instead?

Reply:
Exactly! The relevant issue is [here](https://github.com/paritytech/polkadot-sdk/issues/1638).